src/cswinrt/strings/WinRT.cs

```diff
@@ -415,63 +421,165 @@ protected virtual System.Delegate EventInvoke
                         Expression.Condition(
                             Expression.ReferenceNotEqual(delegateLocal, Expression.Constant(null, typeof(TDelegate))),
                             Expression.Call(delegateLocal, invoke, parameters),
                             Expression.Default(invoke.ReturnType))),
                     parameters).Compile();
 
-                return _eventInvoke;
+                return _state.eventInvoke;
+            }
+        }
+
+        private class Cache
+        {
+            Cache(IWeakReference target, EventSource<TDelegate> source, int index)
+            {
+                this.target = target;
+                SetState(source, index);
+            }
+
+            private IWeakReference target;
+            private ConcurrentDictionary<int, EventSource<TDelegate>.State> states = new ConcurrentDictionary<int, EventSource<TDelegate>.State>();
+
+            private static ReaderWriterLockSlim cachesLock = new ReaderWriterLockSlim();
+            private static ConcurrentDictionary<IntPtr, Cache> caches = new ConcurrentDictionary<IntPtr, Cache>();
+
+            private Cache Update(IWeakReference target, EventSource<TDelegate> source, int index)
+            {
+                // If target no longer exists, destroy cache
+                lock (this)
+                {
+                    using var resolved = this.target.Resolve(typeof(IUnknownVftbl).GUID);
+                    if (resolved == null)
+                    {
+                        this.target = target;
+                        states.Clear();
+                    }
+                }
+                SetState(source, index);
+                return this;
+            }
+
+            private void SetState(EventSource<TDelegate> source, int index)
+            {
+                // If cache exists, use it, else create new
+                if (states.ContainsKey(index))
+                {
+                    source._state = states[index];
+                }
+                else
+                {
+                    source._state = new EventSource<TDelegate>.State();
+                    states[index] = source._state;
+                }
+            }
+
+            public static void Create(IObjectReference obj, EventSource<TDelegate> source, int index)
+            {
+                // If event source implements weak reference support, track event registrations so that
+                // unsubscribes will work across garbage collections.  Note that most static/factory classes
+                // do not implement IWeakReferenceSource, so static codegen caching approach is also used.
+                IWeakReference target = null;
+                try
+                {
+                    target = obj.AsInterface<IWeakReferenceSource>().GetWeakReference();
```

Comment:
AsInterface has a GetCustomAttribute call which we have in the past found to be expensive and worked on reducing.  Can we avoid it in any way like by using AsType instead maybe or maybe an internal version of an optimized AsInterface method which assumes none of the interfaces it gets is for ComImport?